mos-platform/geos-cbm/geos_constants.h

```diff
@@ -0,0 +1,752 @@
+/***********************************************************************
+ * geos_constants.h
+ *
+ * This file contains #defines for use in GEOS applications
+ * with llvm-mos C compiler.
+ *
+ * Transcribed from geoProgrammer manual by DiSc.
+ *
+ ***********************************************************************/
+
+#ifndef GEOS_CONSTANTS_H
+#define GEOS_CONSTANTS_H
+
+/***********************************************************************
+ * Miscellaneous defines
+ ***********************************************************************/
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#define LOW_BYTE(x) (*((unsigned char *)&(##x) + 0))
+#define HIGH_BYTE(x) (*((unsigned char *)&(##x) + 1))
+
+#define WORD(w) (w & 0xff), (w >> 8)
+
+#define AUX(x) #x
+#define STRINGIFY(x) AUX(x)
+
+/***********************************************************************
+ * Hardware related defines
+ ***********************************************************************/
+
+/** values for CPU_DATA - C64 memory config */
+
+#define RAM_64K 0x30        /** 64K RAM */
+#define IO_IN 0x35          /** 60K RAM, 4K I/O space in */
+#define KRNL_IO_IN 0x36     /** kernal and I/O mapped into memory */
+#define KRNL_BAS_IO_IN 0x37 /** kernal, I/O and basic mapped into memory */
+
+/** values for config - C128 mmu */
+
+#define CIOIN 0x7E        /** 60K RAM, 4K I/O space in */
+#define CRAM64K 0x7F      /** 64K RAM */
+#define CKRNLBASIOIN 0x40 /** kernal, I/O and basic ROM's mapped into memory   \
+                           */
+#define CKRNLIOIN 0x4E    /** Kernal ROM and I/O space mapped in */
+
+/***********************************************************************
+ * Menu Equates
+ ***********************************************************************/
+
+/** Menu types */
+typedef enum __attribute__((__packed__)) {
```

Comment:
We default to packed enums in llvm-mos; this should be superfluous.

Reply:
That does not seem to hold true. I put some Debug_Asserts on the needed size of structures and they fail if I remove the attribute from the enums.

Reply:
I set up a little example with this in godbolt, and it looks to be emitting a single-byte enum. Maybe something else is going on?

I'll have to try this on my local install when I get home; it's also possible that godbolt is out of date and that this is a recent regression. We do consider the effects of `-fshort-enums` to be part of the llvm-mos ABI, and that has the effect of placing packed on every enum definition.

See https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:2,endLineNumber:12,positionColumn:2,positionLineNumber:12,selectionStartColumn:2,selectionStartLineNumber:12,startColumn:2,startLineNumber:12),source:'%23include+%3Cstdio.h%3E%0A%0Avolatile+unsigned+s%3B%0Avoid+f(void)+%7B%0A++++enum+%7B%0A++++++HORIZONTAL+%3D+0b00000000,%0A++++++VERTICAL+%3D+0b10000000,%0A++++++CONSTRAINED+%3D+0b01000000,%0A++++++UN_CONSTRAINED+%3D+0b00000000+%0A++++%7D+x%3B%0A++++s+%3D+sizeof(x)%3B%0A%7D'),l:'5',n:'1',o:'C+source+%231',t:'0')),k:57.65137043047197,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cmos-c64-trunk,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-Os',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+llvm-mos+commodore+64+(Editor+%231)',t:'0')),k:42.34862956952804,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4

Reply:
I also quickly ran this example through my local build from tip-of-tree llvm-mos, to the same result.